ppb/features/collision.py

```diff
@@ -0,0 +1,108 @@
+"""
+A collision subsystem with various colliders available.
+"""
+from itertools import product
+
+from ppb.sprites import BaseSprite
+from ppb.systemslib import System
+
+
+class CollidedWithMixin:
+    """
+    Provides a mechanism for CollisionCheckerSystem to inform sprites when
+    they collide with something.
+
+    This is a mixin added to CanCollideCircleMixin and
+    CanCollideSquareMixin. You do not need to to add it to your sprites if
+    you are using one of them.
+    """
+
+    def collided_with(self, other):
+        """
+        Another sprite has collided with this sprite.
+
+        This is meant to be overridden in your sprites. If a sprite has a
+        collider mixin, but doesn't respond to those collisions, you can
+        ignore this.
+        """
+
+
+class CanCollideCircleMixin(BaseSprite, CollidedWithMixin):
+    """
+    Defines a circular region collider and provides a method to check for
+    collision.
+    """
+    radius: float = 1
+
+    def collides_with(self, other: 'CanCollideCircleMixin'):
+        """
+        Check if this sprite collides with another sprite.
+
+        Both sprites must be CanCollideCircleMixin sprites.
+        """
+        distance = (self.position - other.position).length
+        collide_distance = self.radius + other.radius
+        return distance <= collide_distance
+
+
+class CanCollideSquareMixin(BaseSprite, CollidedWithMixin):
+    """
+    Uses the defined square on a sprite to provid a method to check for
+    collision.
+    """
+
+    def collides_with(self, other: 'CanCollideSquareMixin'):
+        """
+        Check if this sprite collides with another sprite.
+
+        Both sprites must be CanCollideSquareMixin sprites.
+        """
+        furthest_left = min(self.left, other.left)
+        furthest_right = max(self.right, other.right)
+        furthest_up = max(self.top, other.top)
+        furthest_down = min(self.bottom, other.bottom)
+        collide_distance = self.size + other.size
+        return all([furthest_right - furthest_left <= collide_distance,
+                    furthest_up - furthest_down <= collide_distance])
+
+
+class CollisionCheckerSystem(System):
+    """
+    A subsystem to check for collisions after the update step of the system.
+
+    To use, add CollisionCheckerSystem to the list of systems in ppb.run
+    or the constructor of ppb.GameEngine.
+
+    You should only use one type of collider with this system. It is not
+    designed to check fallback collision types. It is allowed, however, to
+    have some sprites with colliders and some without.
+    """
+    primed = False
+
+    def on_update(self, event, signal):
+        self.primed = True
+
+    def on_idle(self, event, signal):
+        if not self.primed:
+            return
+        for s1, s2 in product(event.scene, event.scene):
```

Comment:
Leaving a note for myself, I probably want combinations(event.scene, 2) and not product. Should add a test case to prove it.